src/fhirtypes/StructureDefinition.ts

```diff
@@ -133,8 +133,9 @@ export class StructureDefinition {
       if (element.id.startsWith(`${currentId}.`) || element.id.startsWith(`${currentId}:`)) {
         lastMatchId = currentId;
       } else if (
-        !currentId.startsWith(`${lastMatchId}.`) &&
-        !currentId.startsWith(`${lastMatchId}:`)
+        (!currentId.startsWith(`${lastMatchId}.`) && !currentId.startsWith(`${lastMatchId}:`)) ||
+        // If element is not a slice at this level, and the currentId is a slice, break to add children before slices
+        (element.id.startsWith(`${lastMatchId}.`) && currentId.startsWith(`${lastMatchId}:`))
```

Comment:
Put another way, it's a preorder traversal of our element tree, but child nodes always get visited before slice nodes. Excellent!